### PR TITLE
Model-agnostic meta-learning

### LINKED ISSUE #676: Model Agnostic Meta Learning
New ICML paper for low data learning. Most interestingly, has results for regression. We might be able to adapt it to low data molecular regression problems (we only support classification right now). They have a TF implementation.

https://arxiv.org/abs/1703.03400
https://github.com/cbfinn/mam

PRs adapting this code into DeepChem welcome!

### PR DESCRIPTION
**Pull request by peastman**
Fixes #676.  This is my in-progress implementation of MAML.  Comments welcome!

One thing I was especially unsure of was where to put it.  I ended up creating a new top level package called `metalearning`.  Does that seem reasonable?  It just didn't seem to fit anywhere else.

### CODE REVIEW

File: deepchem/metalearning/maml.py [hunk 1]
```diff
@@ -0,0 +1,226 @@
...
+
+class MetaLearner(object):
+  """Model and data to which the MAML algorithm can be applied.
+  
+  To use MAML, create a subclass of this defining the learning problem to solve.
+  It consists of a model that can be trained to perform many different tasks, and
+  data for training it on a large (possibly infinite) set of different tasks.
+  """
+
+  def __init__(self):
```
Comment (rbharath) [inline deepchem/metalearning/maml.py]: Should the explicit `TensorGraph` representation be moved down to a child subgraph? Different networks will probably override this.

Comment (peastman) [inline deepchem/metalearning/maml.py]: I thought I'd deleted that!  Not sure what happened.  Right, the parent class isn't supposed to have a constructor.

**Comment by peastman** [comment]
Something weird happened.  An out of date version of the code somehow got checked in.  It should now be correct.

File: deepchem/metalearning/tests/test_maml.py [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+
+class TestMAML(unittest.TestCase):
+
+  def test_sine(self):
+    """Test meta-learning for sine function."""
+
+    # This is a MetaLearner that learns to generate sine functions with variable
+    # amplitude and phase.
+
+    class SineLearner(dc.metalearning.MetaLearner):
```
Comment (rbharath) [inline deepchem/metalearning/tests/test_maml.py]: The SineLearner would make for a great IPython notebook tutorial!

File: deepchem/metalearning/maml.py [hunk 3]
```diff
@@ -0,0 +1,220 @@
...
+      saver = tf.train.Saver(
+          self.learner.variables, max_to_keep=max_checkpoints_to_keep)
+      checkpoint_index = 0
+      checkpoint_time = time.time()
+
+      # Main optimization loop.
+
+      for i in range(steps):
+        self.learner.select_task()
+        feed_dict = self.learner.get_batch()
```
Comment (peastman) [inline deepchem/metalearning/maml.py]: I'd interpret it as saying that I currently have a hardcoded batch size of 1.  Yes, that secondary level of batching still needs to be implemented.

Comment (lilleswing) [inline deepchem/metalearning/maml.py]: image

**Comment by peastman** [comment]
I've added an example in the low data folder.  If you can look it over and make sure everything I'm doing is reasonable, that would be great.  I'm using toxcast, which has 617 tasks, so I train on 80% of them and validate on the remaining 20%.  Here's the results I'm getting:

```
AUC before fine tuning: 0.5
AUC after fine tuning: 0.801680672269
```

Fine tuning means doing a single step of gradient descent based on a batch of 10 molecules for each task.

**Comment by peastman** [comment]
There's something suspicious going on with that example.  I'm investigating.

**Comment by rbharath** [comment]
After finetuning on a batch of 10 molecules, it would be interesting to print the accuracy on `task_dataset - batch` (all data from the test task except for the batch). This provides a "global measure" of how well the MAML learns that task from limited data. We should probably replicate this experiment with, say, 20, random choices of the finetuning batch.

**Comment by peastman** [comment]
I've updated the example so every batch now includes five positives and five negatives.  With that change, here's the results I'm getting:

```
AUC before fine tuning: 0.844496618106
AUC after fine tuning: 0.846572580645
```

So we're already getting an excellent AUC before doing any fine tuning.  That suggests the tasks are strongly correlated with each other, so the ones we trained on provide lots of information about the ones we validate on.  If that's true, we're not really doing "low data learning".  Let me verify whether that's really true.

**Comment by peastman** [comment]
It turned out those very high AUC scores were coming from the fact that many tasks used identical or nearly identical lists of molecules as their negatives.  Once I randomized them, the scores came down.

But I'm still finding that the AUC barely increases with fine tuning, or occasionally even decreases.  That's despite the fact that other metrics improve a lot.  In addition to AUC, I'm now also calculating the cross entropy loss and the prediction accuracy.  Here's some example numbers for each of those before and after fine tuning:

Cross entropy loss:
before: 13.1193
after: 6.14624

Prediction accuracy:
before: 0.504838709677
after: 0.690322580645

AUC:
before: 0.748251170656
after: 0.744375650364

Does this make sense?  How should I interpret it?  Is AUC an appropriate metric to be using?  Why does it show no improvement when the other two show big improvements?

**Comment by rbharath** [comment]
This is pretty odd to me. I haven't come across a case where there are large accuracy and loss improvements without a corresponding increase in ROC-AUC.

One way to test whether there are bugs might be to try duplicating the image examples from the MAML paper, but with ROC-AUC computed as well. If we can duplicate their accuracy numbers, but see the same flat AUC, it will be a good indication it's something about the underlying algorithm. Else, might be a bug.

**Comment by peastman** [comment]
I pushed a new update to the example.  About half the time when I run it, I get a result like this:

```
Before fine tuning:

Cross entropy loss: 42.5142
Prediction accuracy: 0.5
ROC AUC: 0.700486472425

After fine tuning:

Cross entropy loss: 6.13898
Prediction accuracy: 0.691129032258
ROC AUC: 0.739768470343
```

The loss improves a lot with fine tuning, and the accuracy goes from completely uninformative (50%) to reasonably good.  The AUC, on the other hand, is decent even before fine tuning and only improves a little bit.

On the other hand, the other half of the time I get a result like this:

```
Before fine tuning:

Cross entropy loss: 6.83285
Prediction accuracy: 0.636290322581
ROC AUC: 0.700481269511

After fine tuning:

Cross entropy loss: 6.67773
Prediction accuracy: 0.641935483871
ROC AUC: 0.693076222685
```

All three metrics are reasonably good to begin with, and show little or no improvement with fine tuning.

My interpretation is that toxcast just isn't a good test case for this algorithm.  There's too much correlation between tasks, so it's able to just train a model that works well directly instead of having to train one that's easy to fine tune.  I could try to do a fancier filtering of the data so that each molecule appears as a positive and a negative for equally many tasks, but I'm not sure it's worth the trouble for an example that's really just meant to demonstrate the API.  Perhaps in the future we can find a better data set to use instead.

Anyway, assuming all the tests pass, I think we can call this PR ready to merge.

**Comment by rbharath** [comment]
This seems a reasonable analysis to me.

I definitely agree that we need to design some better test cases for meta learning algorithms. I'll give it some thought.

Awesome work btw @peastman! This is an epic implementation.

LGTM
